text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,676 @@
+bounds for types substituted for `T`.
+
+## Drop
+
+A notable use case of `impl const` is defining `Drop` impls.
+Since const evaluation has no side effects, there is no simple example that
+showcases `const Drop` in any useful way. Instead we create a `Drop` impl that
+has user visible side effects:
+
+```rust
+let x = Cell::new(42);
+SomeDropType(&x);
+// x is now 41
+
+struct SomeDropType<'a>(&'a Cell<u32>);
+impl const Drop for SomeDropType {
+    fn drop(&mut self) {
+        self.0.set(self.0.get() - 1);
+    }
+}
+```
+
+You are now allowed to actually let a value of `SomeDropType` get dropped within a constant
+evaluation. This means that
+
+```rust
+(SomeDropType(&Cell::new(42)), 42).1
+```
+
+is now allowed, because we can prove
+that everything from the creation of the value to the destruction is const evaluable.
+
+Note that one cannot implement `const Drop` for structs with fields with just a regular `Drop` impl.
+While from a language perspective nothing speaks against that, this would be very surprising for
+users. Additionally it would make `const Drop` pretty useless. This is explained in more detail in
+the following subsection about generic code and `const Drop`.
+
+```rust
+struct Foo;
+impl Drop for Foo { fn drop(&mut self) {} }
+struct Bar(Foo);
+impl const Drop for Bar { fn drop(&mut self) {} } // not ok
+// cannot call with `T == Foo`, because of missing `const Drop` impl
+fn foo<T: Drop>(t: T) {
+    // Let t run out of scope and get dropped.
+    // Would not be ok if `T` is `Bar`,
+    // because the drop glue would drop `Bar`'s `Foo` field after the `Bar::drop` had been called.
+    // This function is therefore not accept by the compiler.
+}
+```
+
+### const Drop in generic code
+
+`Drop` is special in Rust. You don't need to specify `T: Drop`, but `T::drop` will still be called
+if an object of type `T` goes out of scope. This means there's an implicit assumption, that given
+an arbitrary `T`, we might call `T::drop` if `T` has a drop impl. While we can specify
+`T: Drop` to allow calling `T::drop` in a `const fn`, this means we can't pass e.g. `u32` for
+`T`, because `u32` has no `Drop` impl. Even types that definitely need dropping, but have no
+explicit `Drop` impl (like `struct Foo(String);`), cannot be passed if `T` requires a `Drop` bound.
+
+To summarize, there are currently three ways to interact with `Drop`:
+
+* don't mention `Drop` in the parameter bounds (or mention `?Drop`, amounting to the same thing)
+    * can pass any type that fulfills the other bounds, but may never go out of scope
+* mention `?const Drop` in the parameter bounds
+    * can only pass types with explicit (const or not) `Drop` impls,
+      still can't drop any values inside the function
+* mention `Drop` in the parameter bounds
+    * can only pass types with explicit `const Drop` impls (so no `u32`)
+
+To resolve this, the language gets a new marker trait `ConstDrop` which is automatically implemented
```

Comment:
nit: I still think I prefer spelling this as `const drop`, since it doesn't look like a normal implementable trait (which it isn't), and I think it more clearly conveys the key point (that `std::mem::drop` for this type is a `const fn`).

Reply:
It is normally implementable, just like `Copy` is normally implementable. We can't infer it. If we inferred it (e.g. like `Send` or `Sync`), adding a private field of a type not implementing `ConstDrop`, that would be a breaking change, just like adding a private field that is `!Send` is a breaking change if all fields were `Send` before.